### PR TITLE
Add Smoke and Foam chemical reaction effects.

### PR DESCRIPTION
**Pull request by Radrark**
Draft because I have not added the foam effect yet, but the smoke one is fully functional:
- It spreads in a range proportional to the reaction size.
- It consumes all the reagents from the solution it was created from.
- It repeatedly applies the touch effect of every reagent it has to every entity it's covering.
- It gradually copies the reagents it has to every bloodstreams of entities it's covering, except those with internals on.

The foam effect will be similar. Any comments or requests are appreciated.

### CODE REVIEW

File: Content.Server/Chemistry/ReactionEffects/SmokeReactionEffect.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            serializer.DataField(ref _rangeMultiplier, "rangeMultiplier",0.2f);
+            serializer.DataField(ref _duration, "duration", 10f);
+            serializer.DataField(ref _spreadDelay, "spreadDelay", 0.5f);
+            serializer.DataField(ref _smokeSound, "smokeSound",string.Empty);
+        }
+
+        public void React(IEntity solutionEntity, double intensity)
+        {
+            var contents = solutionEntity.GetComponent<SolutionContainerComponent>();
```
Comment (collinlunn) [inline Content.Server/Chemistry/ReactionEffects/SmokeReactionEffect.cs:44]: Instead of using GetComponent, which throws an exception if that component is not there, use TryGetComponent, report an error with Logger.Error then return out, so that it doesn't crash the server if something is messed up. Same for GetComponent< SmokeComponent >().

File: Content.Server/Chemistry/ReactionEffects/SmokeReactionEffect.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+            var solution = contents.SplitSolution(contents.MaxVolume);
+            var amount = (int) Math.Round(intensity * _rangeMultiplier);
+
+            var mapManager = IoCManager.Resolve<IMapManager>();
+            if (!mapManager.TryFindGridAt(solutionEntity.Transform.MapPosition, out var grid)) return;
+
+            var coords = grid.MapToGrid(solutionEntity.Transform.MapPosition);
+            var ent = solutionEntity.EntityManager.SpawnEntity("smoke_cloud", coords.SnapToGrid());
```
Comment (collinlunn) [inline Content.Server/Chemistry/ReactionEffects/SmokeReactionEffect.cs:53]: Preferably use a prototype field instead of a specific prototype, so multiple smoke could prototypes can be used.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        private float _averageLifetime;
+
+        public bool IsInception => _inception == Owner;
+
+        public void Activate(Solution solution, int amount, float duration, float spreadDelay)
+        {
+            if (_running)
+                return;
+            TryAddSolution(solution);
+            Start(Owner, amount, duration+2*amount*spreadDelay, spreadDelay, duration+amount*spreadDelay);
```
Comment (collinlunn) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:52]: Its not immediately clear why `lifeTime` and `averageLifeTime` are calculated with those formulas, some explaining comments would help.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            {
+                foreach (var neighbor in snapGrid.GetInDir(dir))
+                {
+                    if (neighbor.TryGetComponent(out SmokeComponent comp) && comp._inception == _inception)
+                        return;
+
+                    if (neighbor.TryGetComponent(out AirtightComponent airtight) && airtight.AirBlocked)
+                        return;
+                }
+                var newSmoke = Owner.EntityManager.SpawnEntity("smoke_cloud", snapGrid.DirectionToGrid(dir));
```
Comment (collinlunn) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:137]: Again would like to not be hardcoded to a specific prototype

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+                    continue;
+
+                if (entity.TryGetComponent(out InternalsComponent internals) &&
+                    internals.BreathToolEntity != null &&
+                    internals.GasTankEntity != null &&
+                    internals.BreathToolEntity.TryGetComponent(out BreathToolComponent breathTool) &&
+                    breathTool.IsFunctional &&
+                    internals.GasTankEntity.TryGetComponent(out GasTankComponent gasTank) &&
+                    gasTank.Air != null)
+                    continue;
```
Comment (collinlunn) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:193]: Add a property to InternalsComponent  that answers "Will this person breath in chems" instead of doing the checks in the SmokeComponent

**Review by collinlunn — COMMENTED** [review]
There are spaces missing after commas in various method calls, please clean those up.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+            foreach (var entity in tile.GetEntitiesInTileFast())
+            {
+                // Reagents touch all entities on the tile
+                foreach (var reagentQuantity in contents.ReagentList.ToArray())
+                {
+                    if (reagentQuantity.Quantity == ReagentUnit.Zero) continue;
+                    var reagent = _prototypeManager.Index<ReagentPrototype>(reagentQuantity.ReagentId);
+                    reagent.ReactionEntity(entity, ReactionMethod.Touch, reagentQuantity.Quantity * solutionFraction);
+                }
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:180]: Going through ReagentList thrice seems redundant, you could probably combine all 3 (at the very least these 2).

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                    internals.GasTankEntity != null &&
+                    internals.BreathToolEntity.TryGetComponent(out BreathToolComponent breathTool) &&
+                    breathTool.IsFunctional &&
+                    internals.GasTankEntity.TryGetComponent(out GasTankComponent gasTank) &&
+                    gasTank.Air != null)
+                    continue;
+
+                var cloneSolution = contents.Solution.Clone();
+                var transferAmount = ReagentUnit.Min(cloneSolution.TotalVolume * solutionFraction, bloodstream.EmptyVolume);
+                var transferSolution = cloneSolution.SplitSolution(transferAmount);
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:197]: Same thing on cloning.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+                foreach (var reagentQuantity in transferSolution.Contents.ToArray())
+                {
+                    if (reagentQuantity.Quantity == ReagentUnit.Zero) continue;
+                    var reagent = _prototypeManager.Index<ReagentPrototype>(reagentQuantity.ReagentId);
+                    transferSolution.RemoveReagent(reagentQuantity.ReagentId,reagent.ReactionEntity(entity, ReactionMethod.Ingestion, reagentQuantity.Quantity));
+                }
+
+                bloodstream.TryTransferSolution(transferSolution);
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:206]: With no cloning you'd need to add it back into the smoke if the transfer failed.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+            var addSolution = solution.SplitSolution(ReagentUnit.Min(solution.TotalVolume,contents.EmptyVolume));
+
+            var result = contents.TryAddSolution(addSolution);
+
+            if (!result)
+                return;
+
+            if (Owner.TryGetComponent(out SpriteComponent sprite))
+                sprite.Color = contents.SubstanceColor;
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:114]: Is this expensive? Using sprite states pushes all of the data over the network and doing this for a lot of entities could be a lot of bandwidth.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+        {
+            _inception = inception;
+            _amount = amount;
+            _lifetime = lifetime;
+            _spreadDelay = spreadDelay;
+            _averageLifetime = averageLifetime;
+            _running = true;
+
+            Owner.SpawnTimer((int) (_spreadDelay*1000), SpreadSmoke);
+        }
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:96]: Just make them properties and have whatever spawns it deal with setting them IMO.

Also roll the timer into the update as mentioned elsewhere.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+                _exposeTimer += frameTime;
+            }
+            else
+            {
+                if (_inception.TryGetComponent(out SmokeComponent smokeComp) &&
+                    smokeComp._exposeTimer > ExposeDelay)
+                {
+                    ReactWithTileAndEntities();
+                }
+            }
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:76]: Rather than chucking all of this into the component IMO you're better off having the system keep track of which components are connected which should also have better perf.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+            if (_running)
+                return;
+            TryAddSolution(solution);
+            Start(Owner, amount, duration+2*amount*spreadDelay, spreadDelay, duration+amount*spreadDelay);
+        }
+
+        public void Update(float frameTime)
+        {
+            if (!_running)
+                return;
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:58]: I doubt this would ever not be true?

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+{
+    [RegisterComponent]
+    public class SmokeComponent : Component
+    {
+        private const float ExposeDelay = 0.5f;
+
+        public override string Name => "Smoke";
+
+        [Dependency] private readonly IMapManager _mapManager = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:32]: If you moved the reaction stuff onto the system you could avoid having to store these references on every smoke component.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+        public void Update(float frameTime)
+        {
+            if (!_running)
+                return;
+
+            if (IsInception)
+            {
+                if (_exposeTimer > ExposeDelay)
+                {
+                    _exposeTimer = 0;
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:64]: When this is moved make it -= frameTime to avoid being tickrate dependent.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
tl;dr move some more stuff onto the system mainly.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs (hunk 10 above)
Comment (moonheart08) [reply Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:96]: IMO this should be kept as it makes refactoring in the future easier

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+                        return;
+                }
+                var newSmoke = Owner.EntityManager.SpawnEntity("smoke_cloud", snapGrid.DirectionToGrid(dir));
+                var smokeComponent = newSmoke.GetComponent<SmokeComponent>();
+
+                if (Owner.TryGetComponent(out SolutionContainerComponent contents))
+                {
+                    var solution = contents.Solution.Clone();
+                    smokeComponent.TryAddSolution(solution);
+                }
```
[the remark this answers is not in the document]
Comment (Radrark) [reply Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:144]: In SS13 the reagents get copied to every smoke cloud. The foam effect is the one that dilutes the reagents.

File: Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs (hunk 7 above)
Comment (Radrark) [reply Content.Server/GameObjects/Components/Chemistry/SmokeComponent.cs:197]: In SS13 the reagents in the smoke don't get consumed, otherwise smoke would be useless against crowded areas.

File: Content.Server/Chemistry/ReactionEffects/SmokeReactionEffect.cs [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+            var smokeComponent = ent.GetComponent<SmokeComponent>();
+            smokeComponent.Activate(solution, amount, _duration, _spreadDelay);
```
[the remark this answers is not in the document]
Comment (Radrark) [reply Content.Server/Chemistry/ReactionEffects/SmokeReactionEffect.cs:56]: But how would I add the solution and set the amount using Initialize? They need get added by the thing that spawns the smoke.

**Comment by Radrark** [comment]
Refactored a bit and did all you said except: 
- Changing the cloning of solutions. It works like that in SS13 and I think is fine.
-  Setting netsync of the sprite component to false. I think it's pointless to make a SmokeVisualizer for only changing the color of the sprite.

**Comment by metalgearsloth** [comment]
> 
> 
> Refactored a bit and did all you said except:
> 
>     * Changing the cloning of solutions. It works like that in SS13 and I think it's fine.
> 
>     * Setting netsync of the sprite component to false. I think it's pointless to make a SmokeVisualizer for only changing the color of the sprite.

Yeah the first feels icky but it is how tg does it so it's okay for now
Setting the sprite once is fine for now but the main issue with it is it will send the entire state over when anything changes.

Rest seems good.

**Comment by metalgearsloth** [comment]
Did you want to merge this now or add in foam first?

**Comment by Radrark** [comment]
I'm gonna add the foam first.

**Comment by Radrark** [comment]
Ok so as I was adding the foam I noticed I was copypasting a lot of code and that can't be good right? So I refactored my code like 3 times and got this. Also I added a smoke visualizer. It should be good now, ready for review.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
I read through the tg foam code to have a better understanding of what's going on as some stuff is unintuitive. Despite the number of comments they're mostly really minor changes and I'm pretty happy with it ingame IMO.

Also should give audio attribution somewhere (in a comment is fine).

File: Resources/Prototypes/Entities/Effects/chemistry_effects.yml [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+  id: FoamedIronMetal
+  name: foamed iron metal
+  parent: BaseFoamedMetal
+  components:
+    - type: Sprite
+      sprite: Effects/foam.rsi
+      state: ironfoam
+    - type: Icon
+      sprite: Effects/foam.rsi
+      state: ironfoam
```
[the remark this answers is not in the document]
Comment (Radrark) [reply Resources/Prototypes/Entities/Effects/chemistry_effects.yml:147]: 1. Yes for the entity spawn panel if someone wants to spawn them.
2. The parent entity has it false.

File: Content.Server/GameObjects/Components/Chemistry/FoamComponent.cs [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+            }
+        }
+
+        private void HandleKillMessage()
+        {
+            if (Owner.TryGetComponent(out AppearanceComponent appearance))
+            {
+                appearance.SetData(FoamVisuals.State, true);
+            }
+            Owner.SpawnTimer(600, () =>
```
[the remark this answers is not in the document]
Comment (Radrark) [reply Content.Server/GameObjects/Components/Chemistry/FoamComponent.cs:129]: To solve this it needs prediction right? I'm not sure if I could do that.

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+                effectComponent.Amount = Amount - 1;
+                Inception.Add(effectComponent);
+
+                spawned.Add(newEffect);
+            }
+
+            SpreadToDir(Direction.North);
+            SpreadToDir(Direction.East);
+            SpreadToDir(Direction.South);
+            SpreadToDir(Direction.West);
```
[the remark this answers is not in the document]
Comment (Radrark) [reply Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:73]: Does that change anything? These 4 calls are made in the same frame right?

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+
+    public class AreaEffectInception
+    {
+        private const float ReactionDelay = 0.5f;
+
+        private readonly HashSet<AreaEffectComponent> _group = new();
+
+        private float _lifeTimer;
+        private float _spreadTimer;
+        private float _reactionTimer;
```
[the remark this answers is not in the document]
Comment (Radrark) [reply Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:105]: And make an entity prototype that has this inception component right?

File: Content.Server/Chemistry/ReactionEffects/FoamReactionEffect.cs [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+
+            var solution = contents.SplitSolution(contents.MaxVolume);
+
+            solution.RemoveSolution(solution.TotalVolume*(1-solutionFraction));
+
+            var mapManager = IoCManager.Resolve<IMapManager>();
+            if (!mapManager.TryFindGridAt(solutionEntity.Transform.MapPosition, out var grid)) return;
+
+            var coords = grid.MapToGrid(solutionEntity.Transform.MapPosition);
+            var ent = solutionEntity.EntityManager.SpawnEntity(_foamPrototypeId, coords.SnapToGrid());
```
[the remark this answers is not in the document]
Comment (Radrark) [reply Content.Server/Chemistry/ReactionEffects/FoamReactionEffect.cs]: I think it's okay to have two different foams overlap.

**Comment by Radrark** [comment]
Addressed all the comments except the ones I replied.

File: Content.Server/Chemistry/ReactionEffects/FoamReactionEffect.cs (hunk 21 above)
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/Chemistry/ReactionEffects/FoamReactionEffect.cs]: Okey dokey.

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs]: Ahh I'm dumb I thought the amount was being adjusted each time. You would think the tiles it spreads to would be the same if it's in a corridor or open space so it is weird.

File: Resources/Prototypes/Entities/Effects/chemistry_effects.yml (hunk 17 above)
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Resources/Prototypes/Entities/Effects/chemistry_effects.yml:147]: Entity spawn panel doesn't show abstract+ icon can be pulled from the sprite now. We're trying to phase out iconcomponent where possible and the panel will also use visualizer defaults.

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+        public override void OnRemove()
+        {
+            base.OnRemove();
+            Inception?.Remove(this);
+        }
+
+    }
+
+    [RegisterComponent]
+    public class AreaEffectInceptionComponent : Component
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:123]: Chuck in its own file now.

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+        private float _spreadTimer;
+        private float _reactionTimer;
+
+        private int _amountCounterSpreading;
+        private int _amountCounterRemoving;
+
+        private float _duration;
+        private float _spreadDelay;
+        private float _removeDelay;
+        private float _averageExposures;
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:141]: Could also add VV to these if you wanted now.

**Review by metalgearsloth — COMMENTED** [review]
Rest LGTM, good work! Just a comment on the spawn panel.

File: Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+                state = stateNode.AsString();
+            }
+
+            _foamDissolve = new Animation {Length = TimeSpan.FromSeconds(delay)};
+            {
+                var flick = new AnimationTrackSpriteFlick();
+                _foamDissolve.AnimationTracks.Add(flick);
+                flick.LayerKey = FoamVisualLayers.Base;
+                flick.KeyFrames.Add(new AnimationTrackSpriteFlick.KeyFrame(state, 0f));
+            }
```
Comment (PaulRitter) [inline Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs:41]: why create a new scope here?

File: Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+
+        public override void OnChangeData(AppearanceComponent component)
+        {
+            base.OnChangeData(component);
+
+            if (component.TryGetData<bool>(FoamVisuals.State, out var state))
+            {
+                if (state)
+                {
+                    var animPlayer = component.Owner.GetComponent<AnimationPlayerComponent>();
```
Comment (PaulRitter) [inline Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs:52]: use trygetcomponent here

File: Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+
+                    if(!animPlayer.HasRunningAnimation(AnimationKey))
+                        animPlayer.Play(_foamDissolve, AnimationKey);
+                }
+            }
+
+            if (component.TryGetData<Color>(FoamVisuals.Color, out var color))
+            {
+                var sprite = component.Owner.GetComponent<ISpriteComponent>();
```
Comment (PaulRitter) [inline Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs:61]: use trygetcomponent here

File: Content.Client/GameObjects/Components/Chemistry/SmokeVisualizer.cs [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public class SmokeVisualizer : AppearanceVisualizer
+    {
+        public override void OnChangeData(AppearanceComponent component)
+        {
+            base.OnChangeData(component);
+
+            if (component.TryGetData<Color>(SmokeVisuals.Color, out var color))
+            {
+                var sprite = component.Owner.GetComponent<ISpriteComponent>();
```
Comment (PaulRitter) [inline Content.Client/GameObjects/Components/Chemistry/SmokeVisualizer.cs:16]: use trygetcomponent here

File: Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+using Robust.Client.GameObjects;
+using Robust.Client.GameObjects.Components.Animations;
+using Robust.Client.Interfaces.GameObjects.Components;
+using Robust.Shared.Maths;
+using Robust.Shared.Utility;
+using YamlDotNet.RepresentationModel;
+
+namespace Content.Client.GameObjects.Components.Chemistry
+{
+    public class FoamVisualizer : AppearanceVisualizer
```
Comment (PaulRitter) [inline Content.Client/GameObjects/Components/Chemistry/FoamVisualizer.cs:14]: add a [UsedImplicitly]

File: Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+                    solutionFraction = amount * (1 - _reagentMaxConcentrationFactor) / _reagentDilutionStart +
+                                       _reagentMaxConcentrationFactor;
+                }
+                solution.RemoveSolution(solution.TotalVolume*(1-solutionFraction));
+            }
+
+            if (!_mapManager.TryFindGridAt(solutionEntity.Transform.MapPosition, out var grid)) return;
+
+            var coords = grid.MapToGrid(solutionEntity.Transform.MapPosition);
+            var ent = solutionEntity.EntityManager.SpawnEntity(_prototypeId, coords.SnapToGrid());
```
Comment (PaulRitter) [inline Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs:103]: default value for _prototypeId is null, so this will just spawn an entity without any components if no prototype is provided :/ either check in ExposeData if a prototype was provided or handle it accordingly here

File: Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs [hunk 30]
```diff
@@ -0,0 +1,132 @@
...
+        private float _duration;
+        /// <summary>
+        /// How many seconds between each spread step.
+        /// </summary>
+        private float _spreadDelay;
+
+        private float _removeDelay;
+
+        private string _prototypeId;
+        private string _sound;
```
Comment (PaulRitter) [inline Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs:46]: more xmldocs would probably be appropriate here

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs [hunk 31]
```diff
@@ -0,0 +1,259 @@
...
+        }
+
+        public void InceptionUpdate(float frameTime)
+        {
+            _group.RemoveWhere(effect => effect.Deleted);
+            if (_group.Count == 0)
+                return;
+
+            // Make every outer square from the group spread
+            if (_amountCounterSpreading != 0)
```
Comment (PaulRitter) [inline Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:169]: nitpick: use `>`

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs (hunk 23 above)
Comment (PaulRitter) [inline Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:141]: some xmldocs pls

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs [hunk 32]
```diff
@@ -0,0 +1,259 @@
...
+            }
+
+            // Make every square from the group react with the tile and entities
+            _reactionTimer += frameTime;
+            if (_reactionTimer > ReactionDelay)
+            {
+                _reactionTimer -= ReactionDelay;
+                foreach (var effect in _group)
+                {
+                    effect.React(_averageExposures);
```
Comment (PaulRitter) [inline Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:216]: check that effect isn't already deleted, since effect.Kill() in line 202 will trigger the such

**Review by PaulRitter — CHANGES_REQUESTED** [review]
smokecomponent & foamcomponent NEED to be generalized & not copy pasted. if you want support on this, @ me in discord (username: /paul)

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs (hunk 32 above)
Comment (Radrark) [reply Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs:216]: effect.Kill() already removes the effect from the group so I don't think this is necessary

File: Resources/Prototypes/Recipes/Reactions/chemicals.yml [hunk 33]
```diff
@@ -57,6 +71,94 @@
...
+    - !type:AreaReactionEffect
+      rangeConstant: 0
+      rangeMultiplier: 1.1 #Range formula: rangeConstant + rangeMultiplier*sqrt(ReactionUnits)
+      maxRange: 10
+      duration: 10
+      spreadDelay: 0.5
+      removeDelay: 0.5
+      diluteReagents: false
+      prototypeId: Smoke
+      sound: /Audio/Effects/smoke.ogg #Taken from https://github.com/tgstation/tgstation/blob/master/sound/effects/smoke.ogg
```
Comment (AJCM-git) [inline Resources/Prototypes/Recipes/Reactions/chemicals.yml:93]: Dont put licenses inside the prototype, there should be an license.txt inside the sound folder, put it in there

**Review by AJCM-git — CHANGES_REQUESTED** [review]

**Comment by ike709** [comment]
Conflicts

File: Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs (hunk 32 above)
Comment (PaulRitter) [reply Content.Server/GameObjects/Components/Chemistry/AreaEffectComponent.cs]: correct

**Review by PaulRitter — CHANGES_REQUESTED** [review]
Short foreword before we get started:
I know this must be painful since technically you already have working code, but sadly working is not equal futureproof. (Trust me i've had to learn that too)
In order for future coders to easily implement new AreaEffects or fix bugs that will eventually pop up, its is crucial for us to write code which uses proper abstraction & inheritance. If we were to copypaste two classes, one would fix a bug in the one class & would have to remember to do it in the other one too. Having our logic in one place only is very important to keep [technical debt](https://en.wikipedia.org/wiki/Technical_debt) down. Anyways, enough said, please forgive me for torturing you with this, here we go:


More concrete Requirements of what i'd like to see:
- Rename `AreaEffectComponent` to `SolutionAreaEffectComponent`, makes more sense imo.
- Make `AreaReactionEffect` abstract.
- defer implementation of lines 143-148 of `AreaReactionEffect` (getting of the areaeffectcomponent) to a abstract method `protected abstract SolutionAreaEffectComponent GetAreaEffectComponent(IEntity entity)`
- Create two inheritors of `AreaReactionEffect`: `FoamAreaReactionEffect` & `SmokeAreaReactionEffect`. Implement `GetAreaEffectComponent` in here, leave it empty for now, we will implement it later, need to rework something first.

- Make `SolutionAreaEffectComponent` abstract. This will make sense later.
- Merge `AreaEffectInceptionComponent` and `SolutionAreaEffectComponent` into one Component. The split doesn't make sense imo and just makes it a bit more confusing
- Merge `FoamComponent` and `SmokeComponent` into `SolutionAreaEffectComponent`, and here is how:
  - Logic from `HandleSpreadMessage` and `TryAddSolution` can just be merge
  - Do NOT merge in the logic of `HandleKillMessage`! Rather define a new abstract method `protected abstract void OnKill()` and call it whenever you'd send the KillMessage
  - Now for the big whoppa: `HandleReactMessage`:
    - Afaict the only difference between the two methods here is the last part, the entity interaction.
    - Therefore, to merge it into `SolutionAreaEffectComponent`, merge the code up to last foreach-loop. now define a new abstract method `protected abstract void ReactWithEntity(IEntity entity)`. call this method in the last foreach loop instead of doing all you logic
- Now create two inheritors of `SolutionAreaEffectComponent`: `FoamSolutionAreaEffectComponent` and `SmokeSolutionAreaEffectComponent`
- implement the abstract methods with the logic we left out earlier when merging it into `SolutionAreaEffectComponent`

- Now we loop back to `FoamAreaReactionEffect` and `SmokeAreaReactionEffect`: Implement `GetAreaEffectComponent` by trying to get the corresponding component (either `FoamSolutionAreaEffectComponent` or `SmokeSolutionAreaEffectComponent`)

Done! Now test it out and @ me on discord cause you eventually found out i made a major error when bikeshedding this problem.

**Comment by Radrark** [comment]
Ok first of all thanks for the mini tutorial it must've taken you some time and it helped me a lot. So, I did all the things except merging AreaEffectInceptionComponent and SolutionAreaEffectComponent because I'm receiving conflicting orders, metalgearsloth told me in an earlier review to make the inception component. So by the silicon policy, I'm able to choose what order to follow and I'll keep the inception component :P, I think it makes things simpler and it uses the already implemented infrastructure.

File: Content.Server/GameObjects/Components/Chemistry/SolutionAreaEffectComponent.cs [hunk 34]
```diff
@@ -0,0 +1,188 @@
...
+    /// </summary>
+    public abstract class SolutionAreaEffectComponent : Component
+    {
+        [Dependency] protected readonly IMapManager MapManager = default!;
+        [Dependency] protected readonly IPrototypeManager PrototypeManager = default!;
+
+        [ComponentDependency] protected readonly SnapGridComponent? SnapGridComponent = default!;
+        [ComponentDependency] protected readonly SolutionContainerComponent? SolutionContainerComponent = default!;
+        public int Amount { get; set; }
+        public SolutionAreaEffectInceptionComponent? Inception { get; set; }
```
Comment (PaulRitter) [inline Content.Server/GameObjects/Components/Chemistry/SolutionAreaEffectComponent.cs:32]: the inceptioncomponent does not need to be an actual component. also the ticking should probably just be moved into this class.

What you should do is:
- Move `InceptionUpdate` into this class
- Replace the component with a `SolutionAreaEffectGroup`-Class that still has the add/remove functions. The newly ported over function then just takes the hashset from inside the solutionareaeffectgroup

File: Content.Server/Chemistry/ReactionEffects/SmokeAreaReactionEffect.cs [hunk 35]
```diff
@@ -0,0 +1,15 @@
...
+using Robust.Shared.Interfaces.GameObjects;
+
+namespace Content.Server.Chemistry.ReactionEffects
+{
+    [UsedImplicitly]
+    public class SmokeAreaReactionEffect : AreaReactionEffect
+    {
+        protected override SolutionAreaEffectComponent GetAreaEffectComponent(IEntity entity)
+        {
+            return !entity.TryGetComponent(out SmokeSolutionAreaEffectComponent comp) ? null : comp;
```
Comment (PaulRitter) [inline Content.Server/Chemistry/ReactionEffects/SmokeAreaReactionEffect.cs:12]: cleaner to use `GetComponentOrNull`

File: Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs [hunk 36]
```diff
@@ -78,7 +78,7 @@ public class AreaReactionEffect : IReactionEffect
         /// </summary>
         private string _sound;
 
-        public AreaReactionEffect()
+        protected AreaReactionEffect()
```
Comment (PaulRitter) [inline Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs]: why do this?

**Review by PaulRitter — CHANGES_REQUESTED** [review]
looking very good! almost there

File: Content.Server/GameObjects/Components/Chemistry/SolutionAreaEffectComponent.cs (hunk 34 above)
Comment (PaulRitter) [reply Content.Server/GameObjects/Components/Chemistry/SolutionAreaEffectComponent.cs:32]: DONT. FORGET WHAT I SAID

File: Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs (hunk 36 above)
Comment (Radrark) [reply Content.Server/Chemistry/ReactionEffects/AreaReactionEffect.cs]: idk the IDE suggested it. Makes sense though because AreaReactionEffect is an abstract class so there's no point in having a public constructor.

**Review by PaulRitter — CHANGES_REQUESTED** [review]
Looks good so far.

Just gotta change it to use nullability, then you're golden for a merge :)

**Comment by Radrark** [comment]
Done.

File: Content.Server/GameObjects/EntitySystems/SolutionAreaEffectSystem.cs [hunk 37]
```diff
@@ -0,0 +1,18 @@
+﻿using Content.Server.GameObjects.Components.Chemistry;
```
Comment (PaulRitter) [inline Content.Server/GameObjects/EntitySystems/SolutionAreaEffectSystem.cs:1]:
```suggestion
#nullable enable
﻿using Content.Server.GameObjects.Components.Chemistry;
```

**Review by PaulRitter — APPROVED** [review]
